crates/matrix-sdk/src/room/mod.rs

```diff
@@ -90,6 +90,64 @@ pub use self::{
     messages::{Messages, MessagesOptions},
 };
 
+type DeduplicatedRequestMap<Key> = Mutex<BTreeMap<Key, Arc<Mutex<Result<(), ()>>>>>;
+
+/// Handler that properly deduplicates requests to the same endpoint, and will
+/// properly report error upwards in case the concurrent request failed.
+pub(crate) struct DeduplicatedRequestHandler<Key> {
+    inflight: DeduplicatedRequestMap<Key>,
+}
+
+impl<Key> Default for DeduplicatedRequestHandler<Key> {
+    fn default() -> Self {
+        Self { inflight: Default::default() }
+    }
+}
+
+impl<Key: Clone + Ord + std::hash::Hash> DeduplicatedRequestHandler<Key> {
+    async fn run<'a, F: Future<Output = Result<()>> + SendOutsideWasm + 'a>(
+        &self,
+        key: Key,
+        code: F,
```

Comment:
The fact it's deduplicating requests is a bit coincidental, in fact: it could deduplicate anything that must be deduplicated based on a unique key. But yeah, I can do the renaming :+1:

Reply:
I don't mind the more general name of the argument if the type is also named more generally 😉